testing/upgradetest/upgrader.go

```diff
@@ -0,0 +1,365 @@
+	}
+	startParsedVersion, err := version.ParseVersion(startVersionInfo.Binary.String())
+	if err != nil {
+		return fmt.Errorf("failed to get parsed start agent build version (%s): %w", startVersionInfo.Binary.String(), err)
+	}
+	endVersionInfo, err := endFixture.ExecVersion(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get end agent build version info: %w", err)
+	}
+
+	if upgradeOpts.preInstallHook != nil {
+		if err := upgradeOpts.preInstallHook(); err != nil {
+			return fmt.Errorf("pre install hook failed: %w", err)
+		}
+	}
+
+	logger.Logf("Installing version %q", startParsedVersion.VersionWithPrerelease())
+
+	// install the start agent
+	var nonInteractiveFlag bool
+	if Version_8_2_0.Less(*startParsedVersion) {
+		nonInteractiveFlag = true
+	}
+	installOpts := atesting.InstallOpts{
+		NonInteractive: nonInteractiveFlag,
+		Force:          true,
+	}
+	output, err := startFixture.Install(ctx, &installOpts)
+	if err != nil {
+		return fmt.Errorf("failed to install start agent (err: %w) [output: %s]", err, string(output))
+	}
+
+	if upgradeOpts.postInstallHook != nil {
+		if err := upgradeOpts.postInstallHook(); err != nil {
+			return fmt.Errorf("post install hook failed: %w", err)
+		}
+	}
+
+	// wait for the agent to be healthy and correct version
+	err = WaitHealthyAndVersion(ctx, startFixture, startVersionInfo.Binary, 2*time.Minute, 10*time.Second, logger)
+	if err != nil {
+		// context added by WaitHealthyAndVersion
+		return err
+	}
+
+	if upgradeOpts.preUpgradeHook != nil {
+		if err := upgradeOpts.preUpgradeHook(); err != nil {
+			return fmt.Errorf("pre upgrade hook failed: %w", err)
+		}
+	}
+
+	logger.Logf("Upgrading from version %q to version %q", startParsedVersion, endVersionInfo.Binary.String())
+
+	upgradeCmdArgs := []string{"upgrade", endVersionInfo.Binary.String()}
+	if upgradeOpts.customPgp == nil {
+		// unless a custom PGP configuration is provided, upgrade is always using --source-uri
+		if upgradeOpts.sourceURI != "" {
+			// specific ---source-uri
+			upgradeCmdArgs = append(upgradeCmdArgs, "--source-uri", upgradeOpts.sourceURI)
+		} else {
+			// --source-uri from the endFixture
+			srcPkg, err := endFixture.SrcPackage(ctx)
+			if err != nil {
+				return fmt.Errorf("failed to get end agent source package path: %w", err)
+			}
+			sourceURI := "file://" + filepath.Dir(srcPkg)
+			upgradeCmdArgs = append(upgradeCmdArgs, "--source-uri", sourceURI)
+		}
+	} else {
+		if upgradeOpts.sourceURI != "" {
+			upgradeCmdArgs = append(upgradeCmdArgs, "--source-uri", upgradeOpts.sourceURI)
+		}
+
+		if len(upgradeOpts.customPgp.PGP) > 0 {
+			upgradeCmdArgs = append(upgradeCmdArgs, "--pgp", upgradeOpts.customPgp.PGP)
+		}
+
+		if len(upgradeOpts.customPgp.PGPUri) > 0 {
+			upgradeCmdArgs = append(upgradeCmdArgs, "--pgp-uri", upgradeOpts.customPgp.PGPUri)
+		}
+
+		if len(upgradeOpts.customPgp.PGPPath) > 0 {
+			upgradeCmdArgs = append(upgradeCmdArgs, "--pgp-path", upgradeOpts.customPgp.PGPPath)
+		}
+	}
```

Comment:
This still doesn't look great (I know it comes from the very specific logic of upgrade tests with PGP).
I would be tempted to separate the customPGP != nil and customPGP == nil at cost of a little duplication for the sourceURI logic but I leave that choice up to you

Reply:
I will have to go with a `*string` for the `sourceURI` to use that logic. That means the GPG tests would need to set `WithSourceURI("")` that way it doesn't use the local `sourceURI`. I can change to that way of operation.

Reply:
I changed to to how I described. See if you like that way better.

Reply:
Looks good now! Thanks! :+1: